### PR TITLE
Bump QT Creator 8

### PR DESCRIPTION
**Pull request by Peter-Levine**
- **dev-qt/qt-creator: add 8.0.1**
Switched from the qmake to cmake buildsystem.  Added support for the Coco code coverage tool. The 'git' USE flag now also builds the gitlab plugin. The 'wayland' flag was added to enforce built-with-use dependency. 'imageviewer' flag is added for conditional dependence on dev-qt/qtsvg.

- **dev-qt/qt-creator: drop 6.0.0 and 9999**
6.0.0 is shadowed by 8.0.1 and 9999 requires QT6.

Closes: https://github.com/gentoo/gentoo/pull/26675
Signed-off-by: Peter Levine <plevine457@gmail.com>

### CODE REVIEW

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [inline dev-qt/qt-creator/qt-creator-8.0.0.ebuild:222]: ...and where is the unbundled dependency then? Without litehtml and without webengine, the documentation viewer will fall back to some very basic built-in viewer which really sucks (last time I checked at least).

File: dev-qt/qt-creator/qt-creator-9999.ebuild
Comment (Pesa) [inline dev-qt/qt-creator/qt-creator-9999.ebuild]: why bother touching this if you're removing the ebuild in the next commit?

**Review by Pesa — CHANGES_REQUESTED** [review]
I only took a very cursory look at the ebuild, but I'm not sure I like adding tons of USE flags, even for the smallest plugins that have no additional deps (keep in mind that plugins can also be enabled/disabled at runtime). I went back and forth on this several times in the past, but honestly I don't see a good reason to not build the `bookmarks` or the `todo` plugin...

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:222]: I don't have a problem leaving it bundled, but seeing as litehtml is its own project and different devs have different levels of tolerance for bundling, I would rather have waited for the OK from you.

File: dev-qt/qt-creator/qt-creator-9999.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-9999.ebuild]: My mistake.  Didn't mean to copy over 9999.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:222]: And about the plugin USE flags.  I have been back and forth on it more than I care to admit.  I would much prefer to get rid of them.  The end user can decide to enable/disable them at runtime.  I'd unconditionally build most plugins, but maybe keep some metaflags like `clang` and `qml`.  Either, `IUSE="clang debug doc qml systemd test tools wayland"`.  or `IUSE="debug doc test tools"`, or `IUSE="debug doc test"`. Let me know.

**Comment by thesamesam** [comment]
Not a user of qt creator at all but @pesa please ping me if/when it's ok to merge if you're unable.

**Comment by Peter-Levine** [comment]
> and there's no telling when QT6 will land in the gentoo repo

Still masked but it apparently just hit the main repo.

**Comment by Peter-Levine** [comment]
@Pesa What I just git-pushed only has `IUSE="clang doc imageviewer qml serialterminal systemd wayland webengine debug designer qt6 test"`.  With regard to plugin related USE flags kept, `designer` depends on `dev-qt/qttools[designer]` and  `serialterminal` depends on `dev-qt/qtserialport`.  They may be too trivial to keep, as per whatever you recommend.  QT6 support is safely added behind the  `qt6` USE flag since it is `use.mask`ed in `profiles/base`.  Building with `USE=-webengine` now builds the bundled qlitehtml.  It pulls in `dev-libs/gumbo` as a dep. though, instead of relying on the bundled one since it's already in the main tree.  It builds the bundled xxd utility because it's a single c file, AFAIKT it is only used at build-time and not installed, and the only in-tree package to install it is `app-editors/vim-core` which is an unreasonable dep for such a small hex dump utility.  All other plugins ant tools are default installed with exception to the cplusplus tools, as they are with every other distro.  The cplusplus tools (cplusplus-mkvisitor, cplusplus-update-frontend, cplusplus-ast2png, and cplusplus-frontend) are never default installed and appear to be IDE development specific.

If you want to revert or remove any use flags let me know.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild]: > and since different devs have different levels of tolerance for bundling, I would rather have waited for the OK from you.

I'm not sure why you'd need an explicit OK from me since I've already made this decision long ago, the current in-tree ebuild already has the bundled litehtml. And generally speaking, Gentoo has always been fairly pragmatic in this regard. Unbundling makes sense only when we can provide a viable alternative, which isn't the case here. Removing features for the sake of unbundling is not the Gentoo way.

**Comment by Pesa** [comment]
I'm not sure why you can't just follow what the existing (6.0) ebuild is doing, where applicable. It seems as if you're writing a new ebuild from scratch rather than doing a version bump. Granted, this is not a trivial bump due to the qmake->cmake switch, but the two build systems have coexisted in qt-creator for years and are almost identical in many aspects.

**Comment by Pesa** [comment]
> @Pesa What I just git-pushed only has `IUSE="clang doc imageviewer qml serialterminal systemd wayland webengine debug designer qt6 test"`.

Now I think you removed too many. I was probably not sufficiently clear before, the decision behind whether something deserves a flag or not depends on 1) the size of the plugin (some users have complained about compilation times), 2) additional dependencies, 3) whether the plugin is popular or niche/specialized. For example, a large plugin that is likely to be used by very few users should be behind a flag, disabled by default, even if it doesn't have any additional build-time deps (one such example is `baremetal` IIRC). Some plugins may look small but they use some QtCreator internal libraries that quickly increase the amount of compiled code. Other flags exist for historical reasons, for instance `qbs`, it used to have a build-time dep but that's no longer needed. These flags can probably be dropped if you want, assuming the plugin has a reasonable size.

> With regard to plugin related USE flags kept, `designer` depends on `dev-qt/qttools[designer]` and `serialterminal` depends on `dev-qt/qtserialport`.

These are fine.

> QT6 support is safely added behind the `qt6` USE flag since it is `use.mask`ed in `profiles/base`.

Regarding Qt6, if it makes things easier, I'd be ok with keeping qtcreator-8.0 Qt5-only, and then quickly adding a Qt6-only 9.0_beta once it becomes available (a couple of months from now).

> Building with `USE=-webengine` now builds the bundled qlitehtml. It pulls in `dev-libs/gumbo` as a dep. though, instead of relying on the bundled one since it's already in the main tree. It builds the bundled xxd utility because it's a single c file, AFAICT it is only used at build-time and not installed, and the only in-tree package to install it is `app-editors/vim-core` which is an unreasonable dep for such a small hex dump utility.

Sounds good.

> All other plugins and tools are default installed with exception to the cplusplus tools, as they are with every other distro

Again, my preference would be to roughly follow what the existing 6.0 ebuild does, unless there's a reason to deviate (such as removing USE flags for small-ish non-niche plugins).

> The cplusplus tools (cplusplus-mkvisitor, cplusplus-update-frontend, cplusplus-ast2png, and cplusplus-frontend) are never default installed and appear to be IDE development specific.

Yeah those tools should not be compiled at all. Nothing new here.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [inline dev-qt/qt-creator/qt-creator-8.0.0.ebuild:197]: this comment doesn't seem to match the code

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [inline dev-qt/qt-creator/qt-creator-8.0.0.ebuild:287]: disagree with installing these, I don't see them in the official packages or anywhere else.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [inline dev-qt/qt-creator/qt-creator-8.0.0.ebuild:279]: extra space after `docinto`

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [inline dev-qt/qt-creator/qt-creator-8.0.0.ebuild:282]: what happened to `docompress -x`?

**Review by Pesa — CHANGES_REQUESTED** [review]

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild]: Sorry.  For some reason I had assumed it was a recent addition and forgot that it was already bundled in 6.0, likely because it wasn't directly referenced in the ebuild.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:25]: We don't.  over-generalization isn't a vice to me but I'll amend it.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:59]: If you mean "Did it depend on the the perfparser tool which was controlled by the perfprofiler flag and is no longer optional in this ebuild iteration?", then yes.  I'll be amending that by going back to the "the way it was in 6.0" approach that you suggested.  If you mean "Doesn't perfparser optionally depend on`app-arch/zstd`?" , then yes,  since 7.0 via automagic.  Which would require introducing yet  another flag, `zstd`, for optional support in a tool which is already already controlled by the `perfprofiler` flag and `sed`ing yet another file, all to support the corner-case of opening a zstd compressed perf data file.  I opted to force a `app-arch/zstd` dependency on the `perfprofiler` flag instead.   perfparser itself, however, depends on `dev-libs/elfutils`.

**Comment by hansfn** [comment]
I just tested the current 8.0.0 ebuild (for Qt Creator 8.0.1) and discovered two problems:

1. I had to add `include(FindPkgConfig)` in the root CMakeLists.txt since `pkg_check_modules` was used in src/libs/qlitehtml/src/CMakeLists.txt (Upstream bug?) I'm using cmake 3.24.0.
2. I had to install gumbo manually as it was a dependency of qlitehtml that wasn't listed in the ebuild.

But now I'm running Qt Creator 8.0.1. Happily so far.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild]: no worries

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Pesa) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:59]: Apologies for the unclear comment. I was referring to elfutils. If it's still only needed by perfparser/perfprofiler then I'd rather keep that optional as a USE flag.

Regarding zstd, I agree that it would be overkill to introduce a separate flag for it. Having it behind `USE=perfprofiler` sounds good to me. That is, assuming it's used only by that plugin, otherwise things get slightly more complicated.

**Comment by Pesa** [comment]
> I just tested the current 8.0.0 ebuild (for Qt Creator 8.0.1) and discovered two problems:

Thanks for testing!

> 2. I had to install gumbo manually as it was a dependency of qlitehtml that wasn't listed in the ebuild.

I'm puzzled by this one, the ebuild does have a dependency on gumbo... are you building with or without USE=webengine?

**Comment by hansfn** [comment]
Sorry, how could I miss gumbo - it's there:
`!webengine? ( dev-libs/gumbo )`
Yes, I'm building without webengine.

I just assumed it wasn't there because I didn't get an error, but I guess it was because I used `ebuild` to build and install/merge the package instead of `emerge`. My bad.

**Comment by Peter-Levine** [comment]
> I'm not sure why you can't just follow what the existing (6.0) ebuild is doing, where applicable. It seems as if you're writing a new ebuild from scratch rather than doing a version bump. Granted, this is not a trivial bump due to the qmake->cmake switch, but the two build systems have coexisted in qt-creator for years and are almost identical in many aspects.

This is the first time I ever examined the CMake files.  The cmake and qmake build files in qtcreator are not mirror images of one another.  There are a number of differences.  As a trivial example,  zstd compression support in perfparser existed, but wasn't addressed, in .pro/.pri.  And with cmake, I can run `ebuild ... configure` and then examine `${BUILD_DIR}/CMakeCache.txt` for configuration flags, or at least the ones that aren't shadowed by the inclusion or absence of others, before I even look at a `CMakeCache.txt`.  I'm not nearly as fluent with qmake so correct me if I'm wrong, but AFAIK there is no qmake equivalent.  `qmake -query` list only standard Gentoo QT_INSTALL_FOO flags.  Which means I'd have to look at every `.pro` and `.pri ` and hope that I've dotted every "i" and crossed every "t".  If you go in to the subdirectory of `src`, you'll see that the qmake files  are gone.  To make sure that I've accounted for all changes between 6.0.0 and 7.0.0 or 8.0.0,  I did start from scratch.  The original intention was afterwards, to go back and compare 6.0.0 to 8.0.0 and adjust the 6.0.0. build with whatever appeared new.  That was the part I neglected and it sort of took on a life of its own. .  I'll amend that.

Having examined the plugins' deps and size, IMHO, bineditor, bookmarks, classview, emacskeys, fakevim, macros, marketplace, and todo should be considered candidates for use flag removal and unconditional building.  I would be for dropping "studiowelcome" entirely.  It appears to be a broken vestige from the split with Qt 3D Studio.  Even in upstream's own extractable build, clicking on an Example leads to a "Failed opening project" error.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:120]: qtcreator has wayland support since 7.0.0.  This enforces a built-with-use dependence.  It's a strict RDEPEND because qtcreator will build fine without `dev-qt/qtbase-${QT6_PV}[egl]` and `dev-qt/qtwayland-${QT6_PV}` but as seen in https://github.com/gentoo/qt/pull/252#issuecomment-1165732612  opening qtcreator in a wayland environment leads to:
```
qt.qpa.wayland: Failed to load client buffer integration: "wayland-egl"
qt.qpa.wayland: Available client buffer integrations: QList()
```

This issue is resolved as soon as those deps are emerged.  The QT5 dep, i.e., `!qt6? ( >=dev-qt/qtgui-${QT5_PV}[wayland] )`, may be unnecessary as there's no KDE6 wayland environment to test whether, if built against QT5 but without `>=dev-qt/qtgui-${QT5_PV}[wayland]`, it would display a similar error.  Either way it acts as a sanity check, if nothing else.

Comment (ionenwks) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:120]: wrt wayland, there's https://bugs.gentoo.org/864509 given will probably need a central way to pull qtwayland as having this on every packages will be kind of silly and qtwayland should depend on whatever it needs, don't think it needs to be handled in qtcreator

Comment (ionenwks) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:120]: >and qtwayland should depend on whatever it needs

Albeit not sure if EGL is always needed, it depends on opengl right now which maybe is lacking?

**Comment by Peter-Levine** [comment]
@Pesa The debugger plugin is implicitly enabled in 6.0.0.  According to its CMakeFiles.txt, it has a dependency on the qmake plugin when `test` is enabled.  Should I add a `debugger` USE flag and add something like `test? (  debugger? ( qmake ) )`  to `REQUIRED_USE` or take away the `qmake` flag and build it unconditionally, along with debugger?

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:41]: It was on the back-burner since I wanted to examine broken testcases and test deps.  I'll be amending it.

File: dev-qt/qt-creator/qt-creator-8.0.0.ebuild
Comment (Peter-Levine) [reply dev-qt/qt-creator/qt-creator-8.0.0.ebuild:282]: I moved it to the package-agnostic folder `/usr/share/qt5-doc` which isn't default compressed, since qtcreator also looks in there for qch files (it's where `dev-qt/qt-docs` installs to).  But I can move it back.  I can also fix it so it looks in `/usr/share/doc/${PF}` instead of `/usr/share/doc/${PN}`.

**Comment by Peter-Levine** [comment]
@Pesa OK.  It's now pretty much based on the IUSE profile of 6.0.0.   `qt6` was removed but, if anyone wants, I could reopen the PR in the qt overlay for 9999 and/or 8.* with `qt6` support.  Tests are working and gtest is unbundled.

`test? ( mcu? ( dev-cpp/gtest:= ) )` dependency was added, on top of the one already in `DEPEND`, since libMcuSupport.so gets installed with symbols that need to be resolved from gtest/gmock at runtime.  If you want me to the omit mcu testing instead, let me know.

There looks to be optional support for `dev-qt/qtquick3d:5` in qml2puppet.  There's no such ebuild for it but I can add it if you want.

`coco? ( lsp )` and `test? ( qbs qmake )` were added to `REQUIRED_USE` based on `PLUGIN_TEST_DEPENDS` tags in the plugins' `CMakeLists.txt`.

qch files  are now correctly installed to `/usr/share/doc/${PF}`.  qtcreator correctly adds them to the Help viewer on my end.

**Comment by Peter-Levine** [comment]
@Pesa There's also [plugin-haskell](https://github.com/qt-creator/plugin-haskell) and [plugin-fossil-scm](https://github.com/qt-creator/plugin-fossil-scm) which I could eventually add support for, either in build or separately, if they're useful enough and you or qt-project accepts maintainer status.

**Comment by Peter-Levine** [comment]
Taking a look at https://github.com/qt-creator/qt-creator/commit/780f89b377ad939de45601169a16942dd57dfaa0 and the [8.0.0 changelog](https://github.com/qt-creator/qt-creator/blob/master/dist/changelog/changes-8.0.0.md?plain=1#L49), it would appear that, although not explicitly needed at build time, clangd functionality depends on `>=clang-14*` at runtime.  I could make the dep a pure `RDEPEND` on top of what's already in `CDEPEND`, but IMHO it greatly simplifies things just to edit the existing clang dependency expression.  And given that they [removed the libclang based code model](https://github.com/qt-creator/qt-creator/blob/master/dist/changelog/changes-8.0.0.md?plain=1#L38) in 8.0.0, it seems to me that clangd is likely not something that should be optional when the `clang` USE flag is set.

**Comment by thesamesam** [comment]
Is this good to go now? The LLVM 13 dep is becoming problematic.

**Comment by thesamesam** [comment]
ping. I'm going to merge it in a few days if no significant objections, in light of the LLVM 13 dep.

**Comment by Pesa** [comment]
sorry, I've been mostly disconnected for the past month or so. Give me a couple of days to take another look, there were a few major changes since my last review.

**Comment by thesamesam** [comment]
No worries at all, just let me know. Thanks!

**Comment by Pesa** [comment]
> Edit: So much appears to depend on `qmake` and `qbs` during `test`, including a few default enabled plugins, that it just makes sense to add `test? ( qbs qmake )` to `REQUIRED_USE`.

Yeah that's fine.

> Edit2: I'm aware that debugger tests are `sed`ed out in 6.0.0, but with cmake the bad ones are omitted by passing `-DWITH_DEBUGGER_DUMPERS=NO`.

Yes, the dumpers tests were disabled in 6.0.0 and all previous versions, they never worked for me.

> There looks to be optional support for dev-qt/qtquick3d:5 in qml2puppet. There's no such ebuild for it but I can add it if you want.

Given that the next version of qtcreator will drop Qt5 altogether, I'd say don't bother.

> coco? ( lsp ) and test? ( qbs qmake ) were added to REQUIRED_USE based on PLUGIN_TEST_DEPENDS tags in plugins' CMakeLists.txt.

OK

> qch files are now correctly installed to /usr/share/doc/${PF}. qtcreator correctly adds them to the Help viewer on my end.

Are you somehow telling qtcreator where to find them? If not, it may appear to be working because this is a `-r0` ebuild so the revision component is omitted from the install path, but it will stop working for `-r1` and higher.

**Comment by Pesa** [comment]
> Taking a look at [qt-creator/qt-creator@780f89b](https://github.com/qt-creator/qt-creator/commit/780f89b377ad939de45601169a16942dd57dfaa0) and the [8.0.0 changelog](https://github.com/qt-creator/qt-creator/blob/master/dist/changelog/changes-8.0.0.md?plain=1#L49), it would appear that, although not explicitly needed at build time, clangd functionality depends on `>=clang-14*` at runtime. I could make the dep a pure `RDEPEND` on top of what's already in `CDEPEND`, but IMHO it greatly simplifies things just to edit the existing clang dependency expression. And given that they [removed the libclang based code model](https://github.com/qt-creator/qt-creator/blob/master/dist/changelog/changes-8.0.0.md?plain=1#L38) in 8.0.0, it seems to me that clangd is likely not something that should be optional when the `clang` USE flag is set.

I'm not following. Your ebuild already depends on clang-14, what other dependency changes are you suggesting?

**Comment by Peter-Levine** [comment]
> I'm not following. Your ebuild already depends on clang-14, what other dependency changes are you suggesting?

Sorry for the miscommunication.  I was announcing [the change I had just previously applied](https://github.com/gentoo/gentoo/compare/f7362e2b4b62c4422a9c6bcfeb0ab9ed91f8fc51..7f92cf64be3b6bd36be285d1d743c5655f5a9098) and my reasoning.

File: dev-qt/qt-creator/qt-creator-8.0.1.ebuild
Comment (Peter-Levine) [inline dev-qt/qt-creator/qt-creator-8.0.1.ebuild:210]: > Are you somehow telling qtcreator where to find them? If not, it may appear to be working because this is a -r0 ebuild so the revision component is omitted from the install path, but it will stop working for -r1 and higher.

Yes,  right here.  I tested the ebuild by temporarily changing `${PF}` to `qt-creator-8.0.1-r2` and can confirm it still shows the documentation in the help viewer.

**Comment by Peter-Levine** [comment]
@Pesa QT Creator 9.0.0 Beta 1 is out now.   I understand if time is very limited but I feel there has been sufficient testing and feedback, both here and in the QT overlay PR, to warrant merging this.

**Comment by stwnt** [comment]
[Version 8.0.2](https://github.com/qt-creator/qt-creator/releases/tag/v8.0.2) is out now. I managed to compile it already with the same ebuild.

**Comment by genbtc** [comment]
Any progress getting this merged properly ? Bump

**Pull request closed without merging**
